### PR TITLE
Preserve historical total progress in RouteProgress updates 

### LINKED ISSUE #1595: Preserve user's total historic progress in RouteProgress updates
Currently, when a user is navigating a route and a re-route event occurs the visual history of the route traveled is removed when the new route is set.

Instead we should merge the new route onto the historic route traveled, so that the user's end-to-end journey is displayed -- the visual history, as well as the percentage traveled since the original origin (not the origin from the re-route location)

### PR DESCRIPTION
**Pull request by danesfeder**
Closes #1595 

This PR adds three new fields to the `RouteProgress`:
* `.totalDistanceTraveled` total distance traveled along all previous routes including the current route’s `RouteProgress.distanceTraveled`.
* `RouteProgress.totalFractionTraveled` the total fraction traveled along all reroutes including the full distance of the current route.
* `RouteProgress.totalDistance` distance traveled along all previous routes including the full distance of the current route.

TODO:
- [x] Update `Navigator`
- [ ] Unit testing 
- [x] Javadoc
- [x] Update `RouteProgress`

cc @akitchen

### CODE REVIEW

**Comment by danesfeder** [comment]
@kevinkreiser I think we are looking pretty good with snapshot 10.  Question though, what format is the duration in?

```
D/NavigationRouteProcessor: **** **** **** **** **** ****
D/NavigationRouteProcessor: Total distance traveled 116.791259765625
D/NavigationRouteProcessor: Total duration traveled 36200.0
D/NavigationRouteProcessor: Total fraction traveled 0.064925425
D/NavigationRouteProcessor: **** **** **** **** **** ****
D/NavigationRouteProcessor: Total distance traveled 124.8524169921875
D/NavigationRouteProcessor: Total duration traveled 39000.0
D/NavigationRouteProcessor: Total fraction traveled 0.069406696
D/NavigationRouteProcessor: **** **** **** **** **** ****
D/NavigationRouteProcessor: Total distance traveled 133.669921875
D/NavigationRouteProcessor: Total duration traveled 42500.0
D/NavigationRouteProcessor: Total fraction traveled 0.07430844
D/NavigationRouteProcessor: **** **** **** **** **** ****
D/NavigationRouteProcessor: Total distance traveled 145.0531005859375
D/NavigationRouteProcessor: Total duration traveled 46500.0
D/NavigationRouteProcessor: Total fraction traveled 0.08063646
```

**Comment by kevinkreiser** [comment]
@danesfeder the units should be milliseconds like just like the standard java apis do for epoch time.

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgress.java [hunk 1]
```diff
@@ -88,11 +96,20 @@ public double durationRemaining() {
     return (1 - fractionTraveled()) * directionsRoute().duration();
   }
 
+
+  /**
+   * Provides the duration traveled along the total route progress.
+   *
+   * @return {@code long} value representing the duration traveled along the total route progress, in unit
```
Comment (devotaaabel) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgress.java]: should say `float`

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgress.java [hunk 2]
```diff
@@ -105,6 +122,16 @@ public float fractionTraveled() {
     return fractionRemaining;
   }
 
+
+  /**
+   * Get the fraction traveled along the total route progress, this is a float value between 0 and 1 and
+   * isn't guaranteed to reach 1 before the user reaches the end of the route.
+   *
+   * @return a float value between 0 and 1 representing the fraction the user has traveled along the
```
Comment (devotaaabel) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgress.java]: Add or remove the `{@code }` for consistency

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteProcessor.java [hunk 3]
```diff
@@ -134,10 +139,40 @@ private void updateIntersections() {
...
 
-  private void addUpcomingStepPoints(RouteProgress.Builder progressBuilder) {
-    if (upcomingStepPoints != null && !upcomingStepPoints.isEmpty()) {
-      progressBuilder.upcomingStepPoints(upcomingStepPoints);
+  private void addTotalTraveled(MapboxNavigator navigator, RouteProgress.Builder progressBuilder) {
+    List<TrackedRoute> history = navigator.retrieveRouteHistory();
+    double totalDistanceTraveled = 0d;
+    float totalDurationTraveled = 0f;
+    float totalFractionTraveled = 0f;
+    double totalDistance = 0f;
```
Comment (devotaaabel) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteProcessor.java]: `0d`? Though should be optional, just make it consistent with `totalDistanceTraveled`

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgress.java [hunk 4]
```diff
@@ -88,11 +96,20 @@ public double durationRemaining() {
...
   }
 
+
+  /**
+   * Provides the duration traveled along the total route progress.
+   *
+   * @return {@code long} value representing the duration traveled along the total route progress, in unit
+   * seconds
+   */
+  public abstract float totalDurationTraveled();
```
Comment (devotaaabel) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/routeprogress/RouteProgress.java]: Is this also going to be used for historical route progress?

**Comment by Guardiola31337** [comment]
Going to close this as stale and leave https://github.com/mapbox/mapbox-navigation-android/issues/1595 opened until we get confirmation from @asinghal22 on if this is something we want to support or not.

cc @zugaldia

**Pull request closed without merging**